relay-sampling/src/evaluation.rs

```diff
@@ -24,26 +30,148 @@ fn pseudo_random_from_uuid(id: Uuid) -> f64 {
+#[derive(Debug)]
+pub struct ReservoirEvaluator {
+    counters: ReservoirCounters,
+    #[cfg(feature = "redis")]
+    redis_pool: Option<Arc<RedisPool>>,
+    #[cfg(feature = "redis")]
+    org_id: Option<u64>,
+}
+
+impl ReservoirEvaluator {
+    /// Constructor for [`ReservoirEvaluator`].
+    pub fn new(counters: ReservoirCounters) -> Self {
+        Self {
+            counters,
+            #[cfg(feature = "redis")]
+            org_id: None,
+            #[cfg(feature = "redis")]
+            redis_pool: None,
+        }
+    }
+
+    /// Sets the Redis pool and organiation ID for the [`ReservoirEvaluator`].
+    ///
+    /// These values are needed to synchronize with Redis.
+    #[cfg(feature = "redis")]
+    pub fn set_redis(mut self, org_id: u64, redis_pool: Arc<RedisPool>) -> Self {
+        self.org_id = Some(org_id);
+        self.redis_pool = Some(redis_pool);
+        self
+    }
+
+    #[allow(unused_variables)]
+    fn redis_count(&self, rule: RuleId, rule_expiry: Option<&DateTime<Utc>>) -> Option<i64> {
+        #[cfg(feature = "redis")]
+        if let (Some(pool), Some(org_id)) = (self.redis_pool.as_ref(), self.org_id) {
+            let key = ReservoirRuleKey::new(org_id, rule);
+
+            let mut redis_client = pool.client().ok()?;
+            let mut redis_connection = redis_client.connection().ok()?;
+
+            if redis_sampling::set_redis_expiry(&mut redis_connection, &key, rule_expiry).is_err() {
+                relay_log::error!("failed to set redis reservoir rule expiry");
+            }
+
+            match redis_sampling::increment_redis_reservoir_count(&mut redis_connection, &key) {
+                Ok(redis_val) => return Some(redis_val),
+                Err(e) => relay_log::error!("failed to increment redis reservoir count: {}", e),
+            }
+        }
+
+        None
+    }
+
+    /// Gets the local count of a reservoir rule. Increments the count if limit isnt reached.
+    fn local_count(&self, rule: RuleId, limit: i64) -> Option<i64> {
+        let mut map_guard = self.counters.lock().ok()?;
+
+        let counter_value = map_guard.entry(rule).or_insert(0);
+        if *counter_value < limit {
+            *counter_value += 1;
+        }
+
+        Some(*counter_value)
+    }
+
+    fn update_counter(&self, rule: RuleId, new_value: i64) {
+        let Ok(mut map_guard) = self.counters.lock() else {
+            relay_log::error!("failed to lock reservoir counter mutex");
+            return;
+        };
+
+        match map_guard.get_mut(&rule) {
+            Some(value) => *value = new_value,
+            // Logging an error because at this point the value should definitively be here.
+            None => relay_log::error!("failed to retrieve counter entry"),
+        }
+    }
+
+    /// Evaluates a reservoir rule, returning `true` if it should be sampled.
+    pub fn evaluate(&self, rule: RuleId, limit: i64, rule_expiry: Option<&DateTime<Utc>>) -> bool {
```

Comment:
This function should satisfy all the performance optimizations mentioned. 

1. it gets the local count and increments it if limit hasn't been reached.
2. fast return if its above the limit.
3. if redis isnt configured, we return early
4. if redis is configured but the received value is less than local value, we avoid locking again
5. only if we receive from redis, and it is higher, do we update the count by locking the mutex again